certbot/compat/os.py

```diff
@@ -29,3 +29,18 @@
 
 # Clean all remaining importables that are not from the core os module.
 del ourselves, std_os, std_sys
+
+
+# Because of the blocking strategy on file handlers on Windows, rename do not behave as expected
+# with POSIX systems: an exception will be raised if dst already exists.
+def rename(*unused_args, **unused_kwargs):  # pylint: disable=function-redefined
+    """Method os.rename() is forbidden"""
+    raise RuntimeError('Usage of os.rename() is forbidden. '
+                       'Use certbot.compat.filesystem.replace() instead.')
+
+
+# To be consistent across Windows and Linux, filesystem.replace should be used.
```

Comment:
I personally think we should remove this method.

If `os.replace` exists, I think it's fine for people to call it as it has the same behavior on both OSes.

If `os.replace` does not exist because we're on Python 2, would we rather have the behavior that `os.replace` does not exist or have it raise this `RuntimeError`? I personally lean toward the former because it acts more like the standard `os` module and `pylint` will warn about there not being a `replace` attribute while it won't warn about this function raising an exception.

Reply:
I would prefer to keep it. It makes this consistent with the purpose of `certbot.compat.os`, that is to raise exceptions on function that cannot be used. Here `os.replace` cannot be used in Certbot, since it will break on Python 2.

Considering your second paragraph, I would evaluate it towards two opinions you expressed in the past:
* first that we are not committed to follow the `os` Python core module API anymore
* second that pylint process cannot be fully trusted, and abandon for instance the idea of linting specifically each forbidden function, in favor of a global lint of `os` import

I think that raising the same exception than in the other functions will give it a lot of visibility. We should not have any `os.replace` in Certbot codebase, like `os.rename`, and so I think it is more coherent.

Reply:
Like everything else, I think there's pros and cons. My approach means that a standard exception will be raised when running the code on Python 2 and our linter has a chance to catch it. Your approach means that the linter will not catch it, but it will always raise the same custom error. I personally think it's debatable which approach is better.

I'm fine keeping this, but if we do, can you update the comment here? I personally don't think it accurately describes the purpose of raising an exception here. `os.replace` is consistent on Windows and Linux.

Reply:
I tried to balance again the two approaches, but when I consider the development workflow, I still really think it makes things clearer.

Let's consider things like this: you cannot use os.rename, because it does not work on windows, you cannot use os.replace because it does not work on python 2. You need to use filesystem.replace.

In the case of os.rename without forbidding the method by an exception, you would get the error from the CI when code is tested I. Appveyor. For os.replace without forbidding, it would be because lint is not happy, or when executing tests in the CI on Python 2.

So two very different errors should deliver the same message: do not use them, use filesystem instead.

Here I think we simplify the development workflow by giving the same behavior (a runtime exception) with an explicit message about what should be used.

If you are still ok with this, I will update the comments.

Reply:
Updated the comment nonetheless.